### PR TITLE
Refactor to allow for different algorithms to work on different archetypes

### PR DESCRIPTION
**Pull request by Nikoleta-v3**
By refactoring I mean packaging the `axelrod-dojo` in such way that
it is easier to contribute to and easier to implement new archetypes and
optimisation algorithms.

Initially, 8256af5, to follow a similar structure to the `archetypes` an `algorithms` folder
has been created. There a file `genetic_algorthm.py` was created that contains the code to implement
a genetic algorithm (which has now been removed from `utils.py`).

Secondly, 097baa8, pso has been implemented in `algorithms`.  `particle_swarm_optimization.py`
contains the code to run a pso algorithm. The pso algorithm needs a `param_class` and `param_args`,
thereafter it uses a objective function to maximise the payoff. 

Multiprocessing was failing for the algorithm and it has been removed. Note that personally 
I am getting an error when trying to multi process `pso_evolve.py` from the `master` branch. 

A `gambler.py` archetype has been created that works with the pso algorithm. In order for an
archetype to work with the pso three function must be implemented:

- `receive_vector`
-  `vector_to_instance`
- `create_vector_bounds`.

Finally, bbf6222, to show how now other archetypes can also use pso. All we need to do is write these three functions for more archetypes. For example, `FSM` archetype has been written to work with the pso.

### CODE REVIEW

**Comment by drvinceknight** [comment]
This looks good to me @Nikoleta-v3, nice work.

@marcharper would you be against us moving the various command line scripts to a `bin` directory? Just to tidy things a bit.

File: src/axelrod_dojo/algorithms/genetic_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        self.outputer = Outputer(output_filename, mode='a')
+        self.size = size
+        self.objective = objective
+        if not bottleneck:
+            self.bottleneck = size // 4
+        else:
+            self.bottleneck = bottleneck
+        if opponents is None:
+            self.opponents_information = [
+                    PlayerInfo(s, {}) for s in axl.short_run_time_strategies]
```
Comment (marcharper) [inline src/axelrod_dojo/algorithms/genetic_algorithm.py:31]: nit: over-indented (and also a few lines just below)

File: src/axelrod_dojo/utils.py [hunk 2]
```diff
@@ -138,16 +134,27 @@ def params(self):
     def crossover(self, other):
         pass
 
+    def receive_vector(self, vector):
```
Comment (marcharper) [inline src/axelrod_dojo/utils.py:137]: Can we have docstrings for the new functions? (I know all the old ones don't have them...)

**Comment by marcharper** [comment]
All the code moves are fine as is a bin directory. The multiprocessing version of PSO worked for me -- I had to manually install version 0.7.0 rather than the pip version but it worked. I'd like it to keep working if possible, or at least leave that option in even if you can't get it to work on your machine.

The particle swarm algorithm is better suited for Gambler and HMM that have continuous parameters (and are "stochastic"). For FSM and Lookerup the evolutionary algorithm makes more sense IMO since one has better control over how the mutation proceeds and the parameters are categorical (typically producing deterministic strategies). The existing HMM script used the evolutionary algorithm out of laziness and I suspected that the PSO algorithm would be more efficient. It's fine for each archetype to be able to use either -- I'm just pointing out that there is a reason to prefer one over the other in many cases.

If you don't mind, some docstrings would be nice, doesn't need to be full numpy style, just an explanation of at least what the newly introduced functions are doing.

**Comment by drvinceknight** [comment]
> All the code moves are fine as is a bin directory. The multiprocessing version of PSO worked for me -- I had to manually install version 0.7.0 rather than the pip version but it worked. I'd like it to keep working if possible, or at least leave that option in even if you can't get it to work on your machine.

Perhaps once this PR is in (with mp with `0.7.0` back in) we open an issue to write a test for the multiprocessing case (perhaps the bug was on @Nikoleta-v3's side?).

> 
>  I'm just pointing out that there is a reason to prefer one over the other in many cases.

:+1:

> 
> If you don't mind, some docstrings would be nice, doesn't need to be full numpy style, just an explanation of at least what the newly introduced functions are doing.

:+1: :)

**Comment by Nikoleta-v3** [comment]
Great, thank you for the feedback! I'll address everything as soon as possible.

File: src/axelrod_dojo/algorithms/genetic_algorithm.py [hunk 3]
```diff
@@ -27,11 +27,11 @@ def __init__(self, params_class, params_args, size, objective, output_filename,
         else:
             self.bottleneck = bottleneck
         if opponents is None:
-            self.opponents_information = [
-                    PlayerInfo(s, {}) for s in axl.short_run_time_strategies]
+            self.opponents_information = [PlayerInfo(s, {}) for s in
```
Comment (marcharper) [inline src/axelrod_dojo/algorithms/genetic_algorithm.py]: This is fine and so was the wrap around, it was just double indented.

File: src/axelrod_dojo/algorithms/particle_swarm_optimization.py [hunk 4]
```diff
@@ -3,18 +3,18 @@
 
 from axelrod_dojo.utils import score_params
 from axelrod_dojo.utils import PlayerInfo
+from multiprocessing import Pool, cpu_count
```
Comment (marcharper) [inline src/axelrod_dojo/algorithms/particle_swarm_optimization.py]: Don't need Pool I think.

File: src/axelrod_dojo/archetypes/fsm.py [hunk 5]
```diff
@@ -139,9 +139,20 @@ def parse_repr(cls, s):
...
 
     def receive_vector(self, vector):
+        """Receives a vector and creates an instance attribute called
+        vector."""
         self.vector = vector
 
     def vector_to_instance(self):
+        """Turns the attribute vector in to a FSM player instance.
+         
+        The vector has three parts.The first is used to define the next state
```
Comment (marcharper) [inline src/axelrod_dojo/archetypes/fsm.py]: space: "parts.The"

**Review by marcharper — APPROVED** [review]

**Review by drvinceknight — APPROVED** [review]

**Pull request merged**